Modules/_abc.c

```diff
@@ -0,0 +1,905 @@
+        goto end;
+    }
+
+    /* 2. Check negative cache; may have to invalidate. */
+    incache = _in_weak_set(impl->_abc_negative_cache, subclass);
+    if (incache < 0) {
+        goto end;
+    }
+    int r = PyObject_RichCompareBool(impl->_abc_negative_cache_version,
+                                     abc_invalidation_counter, Py_LT);
+    assert(r >= 0);  // Both should be PyLong
+    if (r > 0) {
+        /* Invalidate the negative cache. */
+        if (PySet_Clear(impl->_abc_negative_cache) < 0) {
+            goto end;
+        }
+        /* INCREF the new value of cache version,
+           then carefully DECREF the old one. */
+        Py_INCREF(abc_invalidation_counter);
+        Py_SETREF(impl->_abc_negative_cache_version, abc_invalidation_counter);
+    } else if (incache > 0) {
+        result = Py_False;
+        goto end;
+    }
+
+    /* 3. Check the subclass hook. */
+    ok = _PyObject_CallMethodIdObjArgs((PyObject *)self, &PyId___subclasshook__,
+                                       subclass, NULL);
+    if (ok == NULL) {
+        goto end;
+    }
+    if (ok == Py_True) {
+        Py_DECREF(ok);
+        if (_add_to_weak_set(impl->_abc_cache, subclass, 0) < 0) {
+            goto end;
+        }
+        result = Py_True;
+        goto end;
+    }
+    if (ok == Py_False) {
+        Py_DECREF(ok);
+        if (_add_to_weak_set(impl->_abc_negative_cache, subclass, 0) < 0) {
+            goto end;
+        }
+        result = Py_False;
+        goto end;
+    }
+    if (ok != Py_NotImplemented) {
+        Py_DECREF(ok);
+        PyErr_SetString(PyExc_AssertionError, "__subclasshook__ must return either"
+                                              " False, True, or NotImplemented");
+        goto end;
+    }
+    Py_DECREF(ok);
+
+    /* 4. Check if it's a direct subclass. */
+    mro = ((PyTypeObject *)subclass)->tp_mro;
+    assert(PyTuple_Check(mro));
+    for (pos = 0; pos < PyTuple_Size(mro); pos++) {
+        PyObject *mro_item = PyTuple_GetItem(mro, pos);
+        if (mro_item == NULL) {
+            goto end;
+        }
+        if ((PyObject *)self == mro_item) {
+            if (_add_to_weak_set(impl->_abc_cache, subclass, 0) < 0) {
+                goto end;
+            }
+            result = Py_True;
+            goto end;
+        }
+    }
+
+    /* 5. Check if it's a subclass of a registered class (recursive). */
+    pos = 0;
+    Py_hash_t hash;
+    _enter_iter(impl->_abc_registry);
+
+    while (_PySet_NextEntry(impl->_abc_registry->data, &pos, &key, &hash)) {
+        rkey = PyWeakref_GetObject(key);
+        if (rkey == Py_None) {
+            continue;
+        }
+        int r = PyObject_IsSubclass(subclass, rkey);
+        if (r < 0) {
+            _exit_iter(impl->_abc_registry);
+            goto end;
+        }
+        if (r > 0) {
+            if (_add_to_weak_set(impl->_abc_cache, subclass, 0) < 0) {
+                _exit_iter(impl->_abc_registry);
+                goto end;
```

Comment:
@1st1  There is also something I wanted to ask about this code. Here we are doing clean-up after error, but `_exit_iter` itself may fail (return -1). Are there any precautions about such situations? Should we somehow modify exceptions or everything will happen automatically?

Reply:
We can suppress errors from `_exit_iter()`.
https://docs.python.org/3/c-api/exceptions.html#c.PyErr_Fetch

Or we can remove guarded set entirely:

```
registry_list = PySequence_Fast(impl->_registry);  // copy to list.
if (!registry_list) {
    goto end;
}
for (Py_ssize_t i = 0; i < PySequence_Fast_GET_SIZE(registry_list); i++) {
    rkey = PySequence_Fast_GET_ITEM(registry_list, i);
    ...
}
Py_DECREF(registry_list);
```

Guard is required because we can't remove from set while iterating set.
So if we stop iterating from set directly, we can remove guarded_set.

Reply:
That would be a possible improvement if `PySequence_Fast` doesn't itself iterates under the hood, but I am not familiar with it. In any case I would propose to postpone this after beta 1, which comes in few hours.

Reply:
Just iterating set is not problem.  It doesn't call any Python code.

Problem is we call `PyObject_IsSubclass` and `PySet_Add` while iterating set.
It may delete class inheriting ABC and weakref callback may happens.

But I'm OK to postpone after b1.